storage-node/src/cache/data_store_cache/memdisk/mod.rs

```diff
@@ -138,26 +167,73 @@ impl<R: DataStoreReplacer<MemDiskStoreReplacerKey, MemDiskStoreReplacerValue>> D
             }
         }
 
-        let mut disk_replacer = self.disk_replacer.write().await;
-        match disk_replacer.get(&remote_location) {
-            Some(replacer_value) => {
-                let data_store_key = replacer_value.as_value();
-                if let Some(data) = self.disk_store.read_data(data_store_key).await? {
-                    Ok(Some(data))
-                } else {
-                    unreachable!()
+        let data_store_key;
+        {
+            let mut disk_replacer = self.disk_replacer.lock().await;
+            match disk_replacer.get(&remote_location) {
+                Some(replacer_value) => {
+                    data_store_key = replacer_value.as_value().clone();
                 }
+                None => return Ok(None),
             }
-            None => Ok(None),
+        }
+        if let Some(data) = self
+            .disk_store
+            .read_data(&data_store_key, self.disk_replacer.clone())
+            .await?
+        {
+            Ok(Some(data))
+        } else {
+            unreachable!();
         }
     }
 
     async fn put_data_to_cache(
-        &mut self,
+        &self,
         remote_location: String,
         _data_size: Option<usize>,
         mut data_stream: StorageReaderStream,
     ) -> ParpulseResult<usize> {
+        {
+            // If in_progress of remote_location thread fails, it will clean the data from this hash map.
+            let mut existed = false;
+            loop {
+                let (status, size, notify);
+                if let Some(((status_ref, size_ref), notify_ref)) =
+                    self.status_of_keys.read().await.get(&remote_location)
+                {
+                    existed = true;
+                    status = status_ref.clone();
+                    size = *size_ref;
+                    notify = notify_ref.clone();
```

Comment:
Because when the thread wants to `notify.0.notified().await`, it should release the lock of `status_of_keys `. And updating the waiter number should be as close as possible to  `notify.0.notified().await` : (
And!!! `notify.0.notified().await` and `*notify.1.lock().await += 1;` should be atomic, I will leave todo here.

Reply:
Let's solve it later, #57

Reply:
I seems to find a way to solve this... But not sure whether it is correct, see latest commit.
Also, tokio swap threads randomly, or only when they cannot make progress? If it is the later, we can discard this changes, since this concurrency issue seems to never happen.